src/io.rs

```diff
@@ -32,30 +37,181 @@ where
+}
+
+impl fmt::Display for RecordReadError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+/// A type for deserializing TFRecord formats
+#[derive(Debug)]
+pub struct RecordReader<R: Read> {
+    reader: R,
+}
+
+impl<R> RecordReader<R>
+where
+    R: Read,
+{
+    /// Construct a new RecordReader from an underlying Read.
+    pub fn new(reader: R) -> Self {
+        RecordReader { reader }
+    }
+
+    fn read_next_len(&mut self) -> Result<Option<u64>, RecordReadError> {
+        let len = match self.reader.read_u64::<LittleEndian>() {
+            Err(e) => {
+                if e.kind() == io::ErrorKind::UnexpectedEof {
+                    return Ok(None);
+                }
+                return Err(e.into());
+            }
+            Ok(val) => val,
+        };
+
+        let mut len_bytes = [0u8; 8];
+        LittleEndian::write_u64(&mut len_bytes, len);
+
+        let expected_len_crc32 = self.reader.read_u32::<LittleEndian>()?;
+        let actual_len_crc32 = mask(crc32::checksum_castagnoli(&len_bytes));
+        if expected_len_crc32 != actual_len_crc32 {
+            return Err(RecordReadError::InvalidChecksum);
+        }
+        Ok(Some(len))
+    }
+
+    fn checksum_bytes(&mut self, bytes: &[u8]) -> Result<(), RecordReadError> {
+        let actual_bytes_crc32 = mask(crc32::checksum_castagnoli(&bytes));
+        let expected_bytes_crc32 = self.reader.read_u32::<LittleEndian>()?;
+        if actual_bytes_crc32 != expected_bytes_crc32 {
+            return Err(RecordReadError::InvalidChecksum);
+        }
+        Ok(())
+    }
+    fn read_bytes_exact(&mut self, len: u64, buf: &mut [u8]) -> Result<(), RecordReadError> {
+        if (buf.len() as u64) < len {
+            return Err(RecordReadError::BufferTooShort {
+                needed: len,
+                had: buf.len() as u64,
+            });
+        }
+        self.reader.read_exact(buf)?;
+        self.checksum_bytes(buf)?;
+        Ok(())
+    }
+    /// Read the next record into a byte slice.
+    /// Returns the number of bytes read, if successful.
+    /// Returns None, if it could read exactly 0 bytes (indicating EOF)
+    /// let file = std::fs::OpenOptions::new().read(true).open(path).unwrap();
+    /// // Optional, but we probably want to buffer our reads.
+    /// let file = std::io::BufReader::new(file);
+    /// let records = RecordReader::new(file);
+    /// // If we know in advance the maximum length of the records, we can avoid heap allocs
+    /// let mut buf = [ 0u8; 12000 ];
+    /// loop {
+    ///     let next = records.read_next(&mut buf);
+    ///     match next {
+    ///         Ok(res) => match res {
+    ///             Some(len) => train(&buf[0..len]),
+    ///             None => break,
+    ///         }
+    ///         Err(e) => { warn!("{:?}", e); break }
+    ///     }
+    /// }
+    pub fn read_next(&mut self, buf: &mut [u8]) -> Result<Option<u64>, RecordReadError> {
```

Comment:
Good catch.

I went with the following approach:

+ pub fn `peek_next_len()`, which tells the length of the next record, and can be used to resize a heap buffer in advance of the `read_next` call.  Rather than wrap into a `BufRead` internally, I added `Seek` bounds, giving control to the caller. In most cases they will be passing `BufRead`s anyway.
+ `read_next()` skips records which are longer than the supplied buffer. So if you don't know required buffer size, peek first